File: slp/backend-dotnet/Features/Queue/RedisQueueService.cs
using System.Text.Json;
using StackExchange.Redis;

namespace backend_dotnet.Features.Queue;

/// <summary>
/// Redis-backed queue that uses three keys per job:
/// <list type="bullet">
///   <item><c>llm:queue:pending</c>   — list of jobId strings awaiting processing</item>
///   <item><c>llm:queue:processing</c> — list of jobId strings currently in flight</item>
///   <item><c>llm:job:{jobId}</c>     — JSON blob of the full <see cref="LlmJob"/></item>
/// </list>
/// RPOPLPUSH atomically moves a jobId from pending → processing, ensuring no job
/// is lost if the worker crashes between dequeue and acknowledge.
/// </summary>
public class RedisQueueService : IQueueService
{
    private const string PendingKey    = "llm:queue:pending";
    private const string ProcessingKey = "llm:queue:processing";
    private static string JobDataKey(string jobId) => $"llm:job:{jobId}";

    private readonly IConnectionMultiplexer _redis;
    private readonly ILogger<RedisQueueService> _logger;

    public RedisQueueService(IConnectionMultiplexer redis, ILogger<RedisQueueService> logger)
    {
        _redis  = redis;
        _logger = logger;
    }

    /// <inheritdoc/>
    public async Task EnqueueAsync(LlmJob job)
    {
        var db  = _redis.GetDatabase();
        var json = JsonSerializer.Serialize(job);

        // Store full job data — TTL 7 days to self-clean old jobs
        await db.StringSetAsync(JobDataKey(job.JobId), json, TimeSpan.FromDays(7));

        // Push jobId onto left end of pending list
        await db.ListLeftPushAsync(PendingKey, job.JobId);

        _logger.LogDebug("Enqueued job {JobId} ({RequestType})", job.JobId, job.RequestType);
    }

    /// <inheritdoc/>
    public async Task<LlmJob?> DequeueAsync(CancellationToken ct = default)
    {
        var db = _redis.GetDatabase();

        // Atomically move one jobId from right of pending → left of processing
        var jobIdValue = await db.ListRightPopLeftPushAsync(PendingKey, ProcessingKey);
        if (jobIdValue.IsNullOrEmpty)
            return null;

        var jobId = jobIdValue.ToString();

        var jobJson = await db.StringGetAsync(JobDataKey(jobId));
        if (jobJson.IsNullOrEmpty)
        {
            // Orphaned entry — clean up processing queue
            _logger.LogWarning("Job data missing for jobId {JobId}; removing from processing queue", jobId);
            await db.ListRemoveAsync(ProcessingKey, jobIdValue);
            return null;
        }

        return JsonSerializer.Deserialize<LlmJob>(jobJson.ToString());
    }

    /// <inheritdoc/>
    public async Task AcknowledgeAsync(string jobId)
    {
        var db = _redis.GetDatabase();
        var removed = await db.ListRemoveAsync(ProcessingKey, jobId);
        _logger.LogDebug("Acknowledged job {JobId} (removed {Count} entries)", jobId, removed);
    }

    /// <inheritdoc/>
    public async Task<List<string>> GetProcessingJobIdsAsync()
    {
        var db     = _redis.GetDatabase();
        var values = await db.ListRangeAsync(ProcessingKey);
        return values.Select(v => v.ToString()).ToList();
    }

    /// <inheritdoc/>
    public async Task RequeueStaleAsync(string jobId)
    {
        var db = _redis.GetDatabase();

        // Remove from processing list
        await db.ListRemoveAsync(ProcessingKey, jobId);

        // Push back to pending
        await db.ListLeftPushAsync(PendingKey, jobId);

        _logger.LogInformation("Stale job {JobId} moved back to pending queue", jobId);
    }
}

File: slp/backend-nestjs/src/modules/queue/redis-queue.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { IQueueService } from './queue.service.interface';
import { LlmJob } from './llm-job.interface';
import { RedisConnectionFactory } from './redis-connection.factory';

@Injectable()
export class RedisQueueService implements IQueueService {
  private readonly logger = new Logger(RedisQueueService.name);
  private readonly redis;
  private readonly pendingKey = 'llm:queue:pending';
  private readonly processingKey = 'llm:queue:processing';
  private static jobDataKey(jobId: string) {
    return `llm:job:${jobId}`;
  }

  constructor(private redisFactory: RedisConnectionFactory) {
    this.redis = redisFactory.getConnection();
  }

  get isAvailable(): boolean {
    return this.redis !== null;
  }

  async enqueue(job: LlmJob): Promise<void> {
    if (!this.redis) {
      this.logUnavailable();
      return;
    }

    const json = JSON.stringify(job);
    // Store job data with 7-day TTL
    await this.redis.setex(RedisQueueService.jobDataKey(job.jobId), 604800, json);
    // Push to pending list
    await this.redis.lpush(this.pendingKey, job.jobId);
    this.logger.debug(`Enqueued job ${job.jobId}`);
  }

  async dequeue(): Promise<LlmJob | null> {
    if (!this.redis) {
      this.logUnavailable();
      return null;
    }

    // Atomically move jobId from pending → processing (RPOPLPUSH)
    const jobId = await this.redis.rpoplpush(this.pendingKey, this.processingKey);
    if (!jobId) return null;

    const jobJson = await this.redis.get(RedisQueueService.jobDataKey(jobId));
    if (!jobJson) {
      // Orphaned entry – clean up processing list
      await this.redis.lrem(this.processingKey, 0, jobId);
      this.logger.warn(`Job data missing for ${jobId}`);
      return null;
    }

    return JSON.parse(jobJson) as LlmJob;
  }

  async acknowledge(jobId: string): Promise<void> {
    if (!this.redis) {
      this.logUnavailable();
      return;
    }
    await this.redis.lrem(this.processingKey, 0, jobId);
    this.logger.debug(`Acknowledged job ${jobId}`);
  }

  async getProcessingJobIds(): Promise<string[]> {
    if (!this.redis) return [];
    const ids = await this.redis.lrange(this.processingKey, 0, -1);
    return ids;
  }

  async requeueStale(jobId: string): Promise<void> {
    if (!this.redis) return;
    await this.redis.lrem(this.processingKey, 0, jobId);
    await this.redis.lpush(this.pendingKey, jobId);
    this.logger.log(`Requeued stale job ${jobId}`);
  }

  private logUnavailable() {
    // Rate-limit logging? For simplicity, we'll log once per call.
    this.logger.warn('Redis unavailable – queue operation skipped');
  }
}